In [1]:
%load_ext autoreload
%autoreload 2

In [2]:
import pandas as pd

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import warnings
warnings.filterwarnings("ignore")

In [3]:
import sys
sys.path.append("../../../") 

from utils.paths import make_dir_line

modality = 'c'
project = 'Writing Efficient Python Code'
data = make_dir_line(modality, project)

raw = data('raw')

In [4]:
baseball_stats = pd.read_csv(raw / 'baseball_stats.csv', sep = ',', decimal = '.', header = 0, encoding = 'utf-8')

In [5]:
def calc_run_diff(runs_scored, runs_allowed):

    run_diff = runs_scored - runs_allowed

    return run_diff

# 13.4.0 Basic pandas optimizations

## 13.4.2 Iterating with .iterrows()

In [6]:
pit_df = baseball_stats.copy()
pit_df = pit_df[pit_df['Team'] == 'PIT'].reset_index(drop=True)
pit_df = pit_df.loc[:,['Team', 'League', 'Year', 'RS', 'RA', 'W', 'G', 'Playoffs']]
pit_df.head()

Team League  Year   RS   RA   W    G  Playoffs
0  PIT     NL  2012  651  674  79  162         0
1  PIT     NL  2011  610  712  72  162         0
2  PIT     NL  2010  587  866  57  162         0
3  PIT     NL  2009  636  768  62  161         0
4  PIT     NL  2008  735  884  67  162         0

In [7]:
# Iterate over pit_df and print each row
for i, row in pit_df.iterrows():
    print(row)

Team         PIT
League        NL
Year        2012
RS           651
RA           674
W             79
G            162
Playoffs       0
Name: 0, dtype: object
Team         PIT
League        NL
Year        2011
RS           610
RA           712
W             72
G            162
Playoffs       0
Name: 1, dtype: object
Team         PIT
League        NL
Year        2010
RS           587
RA           866
W             57
G            162
Playoffs       0
Name: 2, dtype: object
Team         PIT
League        NL
Year        2009
RS           636
RA           768
W             62
G            161
Playoffs       0
Name: 3, dtype: object
Team         PIT
League        NL
Year        2008
RS           735
RA           884
W             67
G            162
Playoffs       0
Name: 4, dtype: object
Team         PIT
League        NL
Year        2007
RS           724
RA           846
W             68
G            162
Playoffs       0
Name: 5, dtype: object
Team         PIT
League        NL
Year        

In [8]:
# Iterate over pit_df and print each index variable and then each row
for i,row in pit_df.iterrows():
    print(i)
    print(row)
    print(type(row))

0
Team         PIT
League        NL
Year        2012
RS           651
RA           674
W             79
G            162
Playoffs       0
Name: 0, dtype: object
<class 'pandas.core.series.Series'>
1
Team         PIT
League        NL
Year        2011
RS           610
RA           712
W             72
G            162
Playoffs       0
Name: 1, dtype: object
<class 'pandas.core.series.Series'>
2
Team         PIT
League        NL
Year        2010
RS           587
RA           866
W             57
G            162
Playoffs       0
Name: 2, dtype: object
<class 'pandas.core.series.Series'>
3
Team         PIT
League        NL
Year        2009
RS           636
RA           768
W             62
G            161
Playoffs       0
Name: 3, dtype: object
<class 'pandas.core.series.Series'>
4
Team         PIT
League        NL
Year        2008
RS           735
RA           884
W             67
G            162
Playoffs       0
Name: 4, dtype: object
<class 'pandas.core.series.Series'>
5
Team         

In [9]:
# Print the row and type of each row
for row_tuple in pit_df.iterrows():
    print(row_tuple)
    print(type(row_tuple))

(0, Team         PIT
League        NL
Year        2012
RS           651
RA           674
W             79
G            162
Playoffs       0
Name: 0, dtype: object)
<class 'tuple'>
(1, Team         PIT
League        NL
Year        2011
RS           610
RA           712
W             72
G            162
Playoffs       0
Name: 1, dtype: object)
<class 'tuple'>
(2, Team         PIT
League        NL
Year        2010
RS           587
RA           866
W             57
G            162
Playoffs       0
Name: 2, dtype: object)
<class 'tuple'>
(3, Team         PIT
League        NL
Year        2009
RS           636
RA           768
W             62
G            161
Playoffs       0
Name: 3, dtype: object)
<class 'tuple'>
(4, Team         PIT
League        NL
Year        2008
RS           735
RA           884
W             67
G            162
Playoffs       0
Name: 4, dtype: object)
<class 'tuple'>
(5, Team         PIT
League        NL
Year        2007
RS           724
RA           846
W          

## 13.4.3 Run differentials with .iterrows()

In [10]:
giants_df = baseball_stats.copy()
giants_df = giants_df[giants_df['Team'] == 'SFG'].reset_index(drop=True)
giants_df = giants_df.loc[:,['Team', 'League', 'Year', 'RS', 'RA', 'W', 'G', 'Playoffs']]
giants_df.head()

Team League  Year   RS   RA   W    G  Playoffs
0  SFG     NL  2012  718  649  94  162         1
1  SFG     NL  2011  570  578  86  162         0
2  SFG     NL  2010  697  583  92  162         1
3  SFG     NL  2009  657  611  88  162         0
4  SFG     NL  2008  640  759  72  162         0

In [11]:
# Create an empty list to store run differentials
run_diffs = []

# Write a for loop and collect runs allowed and runs scored for each row
for i,row in giants_df.iterrows():
    runs_scored = row['RS']
    runs_allowed = row['RA']
    
    # Use the provided function to calculate run_diff for each row
    run_diff = calc_run_diff(runs_scored, runs_allowed)
    
    # Append each run differential to the output list
    run_diffs.append(run_diff)

giants_df['RD'] = run_diffs
print(giants_df)

   Team League  Year   RS   RA    W    G  Playoffs   RD
0   SFG     NL  2012  718  649   94  162         1   69
1   SFG     NL  2011  570  578   86  162         0   -8
2   SFG     NL  2010  697  583   92  162         1  114
3   SFG     NL  2009  657  611   88  162         0   46
4   SFG     NL  2008  640  759   72  162         0 -119
5   SFG     NL  2007  683  720   71  162         0  -37
6   SFG     NL  2006  746  790   76  161         0  -44
7   SFG     NL  2005  649  745   75  162         0  -96
8   SFG     NL  2004  850  770   91  162         0   80
9   SFG     NL  2003  755  638  100  161         1  117
10  SFG     NL  2002  783  616   95  162         1  167
11  SFG     NL  2001  799  748   90  162         0   51
12  SFG     NL  2000  925  747   97  162         1  178
13  SFG     NL  1999  872  831   86  162         0   41
14  SFG     NL  1998  845  739   89  163         0  106
15  SFG     NL  1997  784  793   90  162         1   -9
16  SFG     NL  1996  752  862   68  162        

## 13.4.5 Iterating with .itertuples()

In [12]:
rangers_df = baseball_stats.copy()
rangers_df = rangers_df[rangers_df['Team'] == 'TEX'].reset_index(drop=True)
rangers_df = rangers_df.loc[:,['Team', 'League', 'Year', 'RS', 'RA', 'W', 'G', 'Playoffs']]
rangers_df.head()

Team League  Year   RS   RA   W    G  Playoffs
0  TEX     AL  2012  808  707  93  162         1
1  TEX     AL  2011  855  677  96  162         1
2  TEX     AL  2010  787  687  90  162         1
3  TEX     AL  2009  784  740  87  162         0
4  TEX     AL  2008  901  967  79  162         0

In [13]:
# Loop over the DataFrame and print each row
for row in rangers_df.itertuples():
  print(row)

Pandas(Index=0, Team='TEX', League='AL', Year=2012, RS=808, RA=707, W=93, G=162, Playoffs=1)
Pandas(Index=1, Team='TEX', League='AL', Year=2011, RS=855, RA=677, W=96, G=162, Playoffs=1)
Pandas(Index=2, Team='TEX', League='AL', Year=2010, RS=787, RA=687, W=90, G=162, Playoffs=1)
Pandas(Index=3, Team='TEX', League='AL', Year=2009, RS=784, RA=740, W=87, G=162, Playoffs=0)
Pandas(Index=4, Team='TEX', League='AL', Year=2008, RS=901, RA=967, W=79, G=162, Playoffs=0)
Pandas(Index=5, Team='TEX', League='AL', Year=2007, RS=816, RA=844, W=75, G=162, Playoffs=0)
Pandas(Index=6, Team='TEX', League='AL', Year=2006, RS=835, RA=784, W=80, G=162, Playoffs=0)
Pandas(Index=7, Team='TEX', League='AL', Year=2005, RS=865, RA=858, W=79, G=162, Playoffs=0)
Pandas(Index=8, Team='TEX', League='AL', Year=2004, RS=860, RA=794, W=89, G=162, Playoffs=0)
Pandas(Index=9, Team='TEX', League='AL', Year=2003, RS=826, RA=969, W=71, G=162, Playoffs=0)
Pandas(Index=10, Team='TEX', League='AL', Year=2002, RS=843, RA=882, W

In [14]:
# Loop over the DataFrame and print each row's Index, Year and Wins (W)
for row in rangers_df.itertuples():
  i = row.Index
  year = row.Year
  wins = row.W
  print(i, year, wins)

0 2012 93
1 2011 96
2 2010 90
3 2009 87
4 2008 79
5 2007 75
6 2006 80
7 2005 79
8 2004 89
9 2003 71
10 2002 72
11 2001 73
12 2000 71
13 1999 95
14 1998 88
15 1997 77
16 1996 90
17 1993 86
18 1992 77
19 1991 85
20 1990 83
21 1989 83
22 1988 70
23 1987 75
24 1986 87
25 1985 62
26 1984 69
27 1983 77
28 1982 64
29 1980 76
30 1979 83
31 1978 87
32 1977 94
33 1976 76
34 1975 79
35 1974 83
36 1973 57


In [15]:
# Loop over the DataFrame and print each row's Index, Year and Wins (W)
for row in rangers_df.itertuples():
  i = row.Index
  year = row.Year
  wins = row.W
  
  # Check if rangers made Playoffs (1 means yes; 0 means no)
  if row.Playoffs == 1:
    print(i, year, wins)

0 2012 93
1 2011 96
2 2010 90
13 1999 95
14 1998 88
16 1996 90


## 13.4.6 Run differentials with .itertuples()

In [16]:
yankees_df = baseball_stats.copy()
yankees_df = yankees_df[yankees_df['Team'] == 'NYY'].reset_index(drop=True)
yankees_df = yankees_df.loc[:,['Team', 'League', 'Year', 'RS', 'RA', 'W', 'G', 'Playoffs']]
yankees_df.head()

Team League  Year   RS   RA    W    G  Playoffs
0  NYY     AL  2012  804  668   95  162         1
1  NYY     AL  2011  867  657   97  162         1
2  NYY     AL  2010  859  693   95  162         1
3  NYY     AL  2009  915  753  103  162         1
4  NYY     AL  2008  789  727   89  162         0

In [17]:
run_diffs = []

# Loop over the DataFrame and calculate each row's run differential
for row in yankees_df.itertuples():
    
    runs_scored = row.RS
    runs_allowed = row.RA

    run_diff = calc_run_diff(runs_scored, runs_allowed)
    
    run_diffs.append(run_diff)

# Append new column
yankees_df['RD'] = run_diffs
print(yankees_df)

   Team League  Year   RS   RA    W    G  Playoffs   RD
0   NYY     AL  2012  804  668   95  162         1  136
1   NYY     AL  2011  867  657   97  162         1  210
2   NYY     AL  2010  859  693   95  162         1  166
3   NYY     AL  2009  915  753  103  162         1  162
4   NYY     AL  2008  789  727   89  162         0   62
5   NYY     AL  2007  968  777   94  162         1  191
6   NYY     AL  2006  930  767   97  162         1  163
7   NYY     AL  2005  886  789   95  162         1   97
8   NYY     AL  2004  897  808  101  162         1   89
9   NYY     AL  2003  877  716  101  163         1  161
10  NYY     AL  2002  897  697  103  161         1  200
11  NYY     AL  2001  804  713   95  161         1   91
12  NYY     AL  2000  871  814   87  161         1   57
13  NYY     AL  1999  900  731   98  162         1  169
14  NYY     AL  1998  965  656  114  162         1  309
15  NYY     AL  1997  891  688   96  162         1  203
16  NYY     AL  1996  871  787   92  162        

Question
In what year within your DataFrame did the New York Yankees have the highest run differential?
You'll need to rerun the code that creates the 'RD' column if you'd like to analyze the DataFrame with code rather than looking at the console output.

R:/ In 1998 (with a Run Differential of 309)

## 13.4.8 Analyzing baseball stats with .apply()

In [19]:
def text_playoffs(num_playoffs): 
    if num_playoffs == 1:
        return 'Yes'
    else:
        return 'No' 

In [18]:
print('Ok_')

Ok_
